nx/lib/nx/defn/sharding_compiler/shard_execution/argument_provider.ex

```diff
@@ -0,0 +1,21 @@
+defmodule Nx.Defn.ShardingCompiler.ShardExecution.ArgumentProvider do
```

Comment:
Why do you need several providers instead of one that receives the index and returns the relevant one?

Reply:
Hrm, this may actually be the simplest, so nevermind.